### PR TITLE
Fix legacy UOM (please consider for 1471, urgent!)

### PR DESCRIPTION
**Pull request by toad**
I am surprised that UOM works in your testing. It looks to me like it shouldn't - we announce via the old message, UOMAnnounce, and it looks like UpdateOverMandatoryManager won't fetch a jar if advertised via the old UOM announcement message?

In any case it would be a very good idea to announce using the new UOMAnnouncement message - that is, please consider the first commit even if you don't want the others, if you can test it.

Please test carefully, I haven't tested this yet. I believe you had a script to automate testing.

### CODE REVIEW

**Comment by Thynix** [comment]
I don't have a script for the test. I have a script for making builds with different update keys / mandatory / not mandatory, and one for inserting builds.

**Comment by Thynix** [comment]
After applying these changes the 1470 node is not able to UOM from a self-mandatory 1471. The 1470 node logs:

    jvm 1    | Node 127.0.0.1:<darknet-port> : Upstream offered us a new main jar (version 1471) but his key was different to ours:
    jvm 1    | our key: USK@nsUNybM0YHldd6n7XPOM0EOQpXkCpe3ALq2QlKMZx2o,mOWcTDRV~KZcFGDe6cUJFiHOjIasbbRtNudLaWyzM2E,AQACAAE/jar/1470
    jvm 1    | his key:USK@ZMKGyKmhkehl7zkZE13zTZyfPWtPHhE7m~s4krxLWjs,0KA8P64dYsGd3Bv9Pm6TEPgOcd7iVGqPZ~rjtxXhcv8,AQACAAE/jar/1471

I can't have Upstream be self-mandatory 1471 and connect to opennet. I could update a "1469" (built from 1470) to a self-mandatory "1470" built from 1471-stable... That's for another day.

**Comment by toad** [comment]
The warning is harmless. It should advertise *both* keys via UOMAnnouncement's, at least once the LegacyJarUpdater has downloaded the blob for 1471-on-old-key. Which should be visible in persistent-temp as legacy-freenet-jar-1421.fblob.

**Comment by Thynix** [comment]
I'm assuming that'll be legacy-freenet-jar-1471.fblob now that the transition version is 1471? Will the fetch complete when the insert was LocalRequestOnly?

**Comment by toad** [comment]
It might be in the node directory, not the persistent-temp. And yes, it should complete. It should hook the keys when the node starts, so it should pick it up if you insert or fetch it, even if it doesn't actually get fetches scheduled. It *should* check the datastore, but priority issues are possible.

File: src/freenet/node/updater/NodeUpdateManager.java [hunk 1]
```diff
@@ -610,20 +587,15 @@ private long canAnnounceUOMNew() {
...
-				mainJarAvailable ? TRANSITION_VERSION : -1,
-				extJarAvailable ? TRANSITION_VERSION_EXT : -1,
-				revocationChecker.lastSucceededDelta(), revocationChecker
-						.getRevocationDNFCounter(), revocationChecker
-						.getBlobSize(),
-				mainJarAvailable ? transitionMainJarFetcher.getBlobSize() : -1,
-				extJarAvailable ? transitionExtJarFetcher.getBlobSize() : -1,
-				(int) node.nodeStats.getNodeAveragePingTime(),
-				(int) node.nodeStats.getBwlimitDelayTime());
+        return DMT.createUOMAnnouncement(updateURI.toString(), revocationURI
```
Comment (Thynix) [inline src/freenet/node/updater/NodeUpdateManager.java]: Oh. `updateURI` should remain `transitionMainJarURIAsUSK`.

**Comment by Thynix** [comment]
After fixing that the upgrade target appears to be attempting to get the update over UOM but it's been ten minutes of it repeatedly logging

> jvm 1    | Fetching main jar from 127.0.0.1:49032 : Upstream
> jvm 1    | Fetching main jar from 192.168.1.83:49032 : Upstream

and the useralert

> It seems that your node isn't running the latest version of the software. Your node will automatically restart as soon as it has finished downloading and verifying the new version of Freenet.

**Comment by Thynix** [comment]
The upgrade target has started logging

> The update process seems to have been stuck for too long; let's switch to UoM! SHOULD NOT HAPPEN! (2) (ext)

**Comment by Thynix** [comment]
With these changes the upgrade target repeatedly logs

> jvm 1    | Node 192.168.1.83:49032 : Upstream offered us a new main jar (version 1471) but his key was different to ours:
> jvm 1    | our key: USK@h2nulovzS--ReMXjOfxXiTmL208z10r7jbNlxWVF8vg,w795NASc-A67F3TFnznrrM5NqshB0-i3sf-n5SHxqxc,AQACAAE/jar/1470
> jvm 1    | his key:USK@kwBKPkT8dbhyIqhJ0j~SNDB6GmActV87hZ9KVHxD0N0,A9e3WADiVWCbZVPFl-PEj~uyM8xu4Sp2ZLu5tI7~orI,AQACAAE/jar/1471
> jvm 1    | Receiving main jar 1471 from 192.168.1.83:49032 : Upstream
> jvm 1    | Failed to fetch main jar 1471 from blob from 192.168.1.83:49032 : Upstream

It's getting both announcements and creating `main-<numbers>.fblob.tmp` in `persistent-temp/`, but fails to extract the jar:

> 3455455-Jan 30, 2016 19:41:52:524 (freenet.client.async.ClientGetter, Datastore checker(11), MINOR): onFailure(FetchException:Data not found:null:-1:null:false:null:null, freenet.client.async.SingleFileFetcher@4747faff) on freenet.client.async.ClientGetter@77ab7edb for USK@h2nulovzS--ReMXjOfxXiTmL208z10r7jbNlxWVF8vg,w795NASc-A67F3TFnznrrM5NqshB0-i3sf-n5SHxqxc,AQACAAE/jar/1471
> 3455456-FetchException:Data not found:null:-1:null:false:null:null
> 3455457-	at freenet.client.async.SimpleSingleFileFetcher.notFoundInStore(SimpleSingleFileFetcher.java:165)
> 3455458-	at freenet.client.async.BaseSingleFileFetcher.preRegister(BaseSingleFileFetcher.java:307)
> 3455459-	at freenet.client.async.ClientRequestScheduler.finishRegister(ClientRequestScheduler.java:192)
> 3455460-	at freenet.client.async.DatastoreChecker.realRun(DatastoreChecker.java:216)
> 3455461-	at freenet.client.async.DatastoreChecker.run(DatastoreChecker.java:112)
> 3455462-	at freenet.support.PooledExecutor$MyThread.innerRun(PooledExecutor.java:248)
> 3455463-	at freenet.support.PooledExecutor$MyThread.realRun(PooledExecutor.java:188)
> 3455464-	at freenet.support.io.NativeThread.run(NativeThread.java:129)
> 3455465-Jan 30, 2016 19:41:52:524 (freenet.client.async.SimpleSingleFileFetcher, <noname>(13), MINOR): onFailure( FetchException:Route not found - could not find enough nodes to be sure the data doesn't exist:null:-1:null:false:null:null , false)
> 3455466-FetchException:Route not found - could not find enough nodes to be sure the data doesn't exist:null:-1:null:false:null:null
> 3455467-	at freenet.node.SendableGet.translateException(SendableGet.java:121)
> 3455468-	at freenet.client.async.SimpleSingleFileFetcher.onFailure(SimpleSingleFileFetcher.java:67)
> 3455469-	at freenet.client.async.ChosenBlockImpl$3.run(ChosenBlockImpl.java:97)
> 3455470-	at freenet.support.DummyJobRunner$1.run(DummyJobRunner.java:35)
> 3455471-	at freenet.support.PooledExecutor$MyThread.innerRun(PooledExecutor.java:248)
> 3455472-	at freenet.support.PooledExecutor$MyThread.realRun(PooledExecutor.java:188)
> 3455473-	at freenet.support.io.NativeThread.run(NativeThread.java:129)
> 3455474:Jan 30, 2016 19:41:52:524 (freenet.node.updater.UpdateOverMandatoryManager$13, Datastore checker(11), ERROR): Failed to fetch main jar 1471 from blob from 192.168.1.83:49032 : Upstream

**Comment by toad** [comment]
Closed in favour of #497

**Pull request closed without merging**